0015-ethereum-htlc-graceful-exit.adoc

```diff
@@ -0,0 +1,310 @@
+- https://github.com/ethereum/EIPs/blob/master/EIPS/eip-140.md[EIP-140: REVERT instruction]
+- https://ethereum.stackexchange.com/questions/119/what-opcodes-are-available-for-the-ethereum-evm
+- https://ethereum.github.io/yellowpaper/paper.pdf[Ethereum Yellow Paper]
+- https://github.com/ethereum/go-ethereum/blob/master/core/vm/opcodes.go
+- https://solidity.readthedocs.io/en/v0.5.11/assembly.html
+
+
+== Research
+
+=== Instructions to terminate a call
+
+==== The `return` instruction
+
+The `return` instruction is the standard `exit` of an EVM program.
+It halts execution, the state remains as it is (ie, it is saved) and data can be returned.
+
+Because the COMIT Ether and ERC20 Basic Swap HTLCs do not have a store or a state that can be changed, the fact that the `return` instruction saves the state does not matter.
+
+In these HTLCs, the `return` instruction is never used to return data, the call is always `return(0,0)`.
+
+Note that if `return` instruction is executed, the transaction status is successful.
+We already determined that it is confusing for the user if the redeem or refund actually failed.
+
+==== The `revert` instruction
+
+As per https://github.com/ethereum/EIPs/blob/master/EIPS/eip-140.md[EIP-140]:
+
+> The REVERT instruction provides a way to stop execution and revert state changes, without consuming all provided gas and with the ability to return a reason.
+
+Which means that the REVERT instructions gives us the benefits we are seeking:
+
+- Stop the execution: _The effect of REVERT is that execution is aborted, *considered as failed*, and state changes are rolled back._
+- Allow an error message to be returned: _The error message will be available to the caller in the returndata buffer and will also be copied to the output area, i.e. it is handled in the same way as the regular return data is handled._
+
+==== Other instructions?
+
+As far as I can see, there are no other instructions that we can use to terminate a call without an undesirable side effect.
+
+The list of closure instructions (`0xf0` range):
+
+----
+0xf0    CREATE          Create a new account with associated code
+0xf1    CALL            Message-call into an account
+0xf2    CALLCODE        Message-call into this account with alternative account's code
+0xf3    RETURN          Halt execution returning output data
+0xf4    DELEGATECALL    Message-call into this account with an alternative account's code, but persisting the current values for `sender` and `value`
+0xfd    REVERT          Halt execution, revert state changes and return output data
+0xff    SELFDESTRUCT    Halt execution and register account for later deletion
+----
+
+
+
+=== Improving logging
+
+Whether we use `return` or `revert`, we could improve the logging by returning or `log`-ging data at the end of a failed execution.
+
+We need to consider whether to
+
+- Use the same message for A, B & C.
```

Comment:
:upside_down_face: Slightly confusing to talk about `A`, `B` and `C` before they're introduced. Also the list that introduces them below is actually rendered as a numbered list so it looks like they're never introduced :stuck_out_tongue:

Reply:
![Aug28::163005](https://user-images.githubusercontent.com/9418575/63831085-2c3b7800-c9b1-11e9-8aee-fff6d94b6e28.png)

Reply:
haha yeah I re-ordered few things there and there. thanks.